### ISSUE TITLE
User-defined graphics library

### ISSUE THREAD

**Issue by Bestduan** [hediet/vscode-drawio#182]
This plug-in is really great, will you add the function of user-defined graphics library in the future? I think it is of great help to the graphical and process-based design of FPGA, looking forward to your reply.

**Comment by hediet** [comment]
It already is supported! See the settings, you can set a path to custom libraries.

**Comment by Bestduan** [comment]
I've found this setting, but I haven't found a specific way to make a custom library, such as how to define graphics properties.

**Comment by hediet** [comment]
Can you elaborate what you mean with "custom library"?
Can you describe exactly what is missing and you would like to see?

**Comment by Bestduan** [comment]
![image](https://user-images.githubusercontent.com/38045864/99188638-ec200b00-2797-11eb-9281-782cd39595dd.png)
In the figure , 1 point to how the graph is called, 2 to how the library is added, and 3 to the properties of the graph being called. Now, I want to customize what the graph looks like and what the properties are.

**Comment by hediet** [comment]
How do you add it in "More Shapes..."? Is `Electrical / Logic Gates` a predefined library or did you import it from some file? Am I right that you want to customize this library?

Is this feature supported by Draw.io (e.g. Draw.io Desktop)?
If not, it will be hard to implement this here.

**Comment by Bestduan** [comment]
![image](https://user-images.githubusercontent.com/38045864/99189105-85502100-279a-11eb-8c82-e86048f8b04d.png)
I've designed a processing module, and now I want to put it in a custom library for the next call.At the same time, this module needs to configure parameters. I want to set the parameters of this module in the property on the right, so that the code can be generated automatically according to the wiring of the graphic property after the wiring is completed.

**Comment by Bestduan** [comment]
> 如何在“更多形状...”中添加它？是`Electrical / Logic Gates`预定义的库还是您从某个文件导入了它？您要自定义此库，对吗？
> 
> Draw.io（例如Draw.io Desktop）是否支持此功能？
> 如果没有，将很难在这里实现。
Yes, I would like to customize a user library similar to Electrical/Logic Gates. I do not known whether this feature is now supported by the plug-in

**Comment by Bestduan** [comment]
> How do you add it in "More Shapes..."? Is `Electrical / Logic Gates` a predefined library or did you import it from some file? Am I right that you want to customize this library?
> 
> Is this feature supported by Draw.io (e.g. Draw.io Desktop)?
> If not, it will be hard to implement this here.

Yes, I would like to customize a user library similar to Electrical/Logic Gates. I do not known whether this feature is now supported by the plug-in

**Comment by nopeslide** [comment]
> How do you add it in "More Shapes..."? Is `Electrical / Logic Gates` a predefined library or did you import it from some file?

It's a predefined one.

> Yes, I would like to customize a user library similar to Electrical/Logic Gates. I do not known whether this feature is now supported by the plug-in

1. You can import any exported library to the scratchpad if you click on the pencil icon in the scratchpad:
![image](https://user-images.githubusercontent.com/42803409/99200728-4bd1e280-27a7-11eb-861f-00e00749b9fd.png)

2. Or add it via `File/Import`

3. Otherwise it should be possible to always load an exported library afaik via
```json
"hediet.vscode-drawio.customLibraries": [
    {"file": "/path/to/lib.xml","libName": "test"}
],
```
but it wasn't added to the libraries on the left side or was reachable the `+ More Shapes` dialog.
@hediet maybe a bug?


> I've designed a processing module, and now I want to put it in a custom library for the next call.At the same time, this module needs to configure parameters. I want to set the parameters of this module in the property on the right, so that the code can be generated automatically according to the wiring of the graphic property after the wiring is completed.

You could use properties as [placeholders](https://desk.draw.io/support/solutions/articles/16000051979-how-to-work-with-placeholders-) (not the custom ones on the right, but the ones reachable via `Strg+M` or `Right click/Edit Data`).
The custom properties on the right are hardcoded into the shape types and can not be modified afaik.

Except if you code your own shapes:
I built an [RTL extension](https://github.com/nopeslide/drawio_rtl_plugin/blob/master/vscode/README.md) for building such block diagrams.

**Comment by hediet** [comment]
How would you create a custom library?

**Comment by nopeslide** [comment]
> How would you create a custom library?

[use-custom-libraries](https://desk.draw.io/support/solutions/articles/16000092763-use-custom-libraries-in-confluence-cloud)
[work-with-custom-shape-libraries](https://desk.draw.io/support/solutions/articles/16000067790-work-with-custom-shape-libraries)

**Comment by Bestduan** [comment]
> > How do you add it in "More Shapes..."? Is `Electrical / Logic Gates` a predefined library or did you import it from some file?
> 
> It's a predefined one.
> 
> > Yes, I would like to customize a user library similar to Electrical/Logic Gates. I do not known whether this feature is now supported by the plug-in
> 
> 1. You can import any exported library to the scratchpad if you click on the pencil icon in the scratchpad:
>    ![image](https://user-images.githubusercontent.com/42803409/99200728-4bd1e280-27a7-11eb-861f-00e00749b9fd.png)
> 2. Or add it via `File/Import`
> 3. Otherwise it should be possible to always load an exported library afaik via settings:
> 
> ```json
> "hediet.vscode-drawio.customLibraries": [
>     {"file": "/path/to/lib.xml","libName": "test"}
> ],
> ```
> 
> but in my case it wasn't added to the libraries on the left side or was reachable in the `+ More Shapes` dialog.
> @hediet maybe a bug?
> 
> > I've designed a processing module, and now I want to put it in a custom library for the next call.At the same time, this module needs to configure parameters. I want to set the parameters of this module in the property on the right, so that the code can be generated automatically according to the wiring of the graphic property after the wiring is completed.
> 
> You could use properties as [placeholders](https://desk.draw.io/support/solutions/articles/16000051979-how-to-work-with-placeholders-) (not the custom ones on the right, but the ones reachable via `Strg+M` or `Right click/Edit Data`).
> The custom properties on the right are hardcoded into the shape types and can not be modified afaik.
> 
> Except if you code your own shapes:
> I built an [RTL extension](https://github.com/nopeslide/drawio_rtl_plugin/blob/master/vscode/README.md) for building such block diagrams.

Thank you for your reply. I will try it

**Comment by Bestduan** [comment]
> > How would you create a custom library?
> 
> [use-custom-libraries](https://desk.draw.io/support/solutions/articles/16000092763-use-custom-libraries-in-confluence-cloud)
> [work-with-custom-shape-libraries](https://desk.draw.io/support/solutions/articles/16000067790-work-with-custom-shape-libraries)

Whether the exported XML file can be used in VSCode？

**Comment by hediet** [comment]
> Whether the exported XML file can be used in VSCode？

You can most probably use the exported xml file by referencing it in the "custom libraries" setting!

**Comment by jonathanwiesel** [comment]
I created a custom library using the standard draw.io webapp then exported it but when importing it using the settings it renders an empty section:

<img width="215" alt="Screenshot 2023-06-05 at 14 36 31" src="https://github.com/hediet/vscode-drawio/assets/1326781/f0616c20-3daa-4e0a-b36d-0691267dad7f">


```
"hediet.vscode-drawio.customLibraries": [
    {"file": "/Users/my_user/testlib.xml", "libName": "test", "entryId": "test"}
]
```

Am I missing something?

**Comment by miz-online** [comment]
I was confused as well.
The issue seems to be, that the extension only accepts compressed XML content, but draw.io only saves libraries in clear text (not sure why, at one point I read that they switched compressed xml off because of the many problems it created).

You can convert the library xml:
* copy the xml code ('xml' property content of the JSON within the XML tag)
* string-unescape (to remove string escaping for quotes)
* copy unescaped string here https://jgraph.github.io/drawio-tools/tools/convert.html
* press encode
* copy the compressed code into the 'xml' property

now the library should load.

Of course it would be way easier, if the extension would allow compressed or uncomrpressed xml content in libraries. Would this be possible?

**Comment by adlerzei** [comment]
This is still an issue. Is there any chance it will be fixed?
